ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -562,27 +563,37 @@ class CPreOrderTraversalGenerator : public ITraveralGenerator
 public:
 	traversal_t genTraversal(const CElementBSDF* _bsdf, IBSDFDataBufOffsetGetter* _getter) override
 	{
+		constexpr uint32_t INVALID_INDEX = 0xdeadbeefu;
+
 		traversal_t traversal;
 		traversal.push_back(static_cast<bsdf::instr_t>(bsdf::OP_SET_GEOM_NORMAL));
-		push(_bsdf, static_cast<bsdf::instr_t>(0), nullptr);
+		push(_bsdf, static_cast<bsdf::instr_t>(0), nullptr, INVALID_INDEX);
 
 		while (!stack.empty())
 		{
 			auto& top = stack.top();
 			const bsdf::E_OPCODE op = bsdf::getOpcode(top.instr);
 
 			const uint32_t bsdfBufIx = _getter->getOffset(top.bsdf, 0u, top.bsdf->type == CElementBSDF::Type::MIXTURE_BSDF ? top.bsdf->mixturebsdf.weights[top.weight_ix] : 0.f, top.maskParent);
+			const uint32_t currIx = traversal.size();
 			traversal.push_back(emitInstr(top.instr, top.bsdf, bsdfBufIx, top.maskParent));
+			if (top.parentIx!=INVALID_INDEX)//means it's right child of a parent
+			{
+				//write RIGHT_JUMP bitfield into "parent instruction"
+				bsdf::instr_t& parentInstr = traversal[top.parentIx];
+				//writing absolute offset, but could be relative (i.e. `currIx-top.parentIx`) as well (decision needed)
+				parentInstr |= static_cast<bsdf::instr_t>(currIx)<<bsdf::INSTR_RIGHT_JUMP_SHIFT;
+			}
```

Comment:
I thought we agreed to have the Light and BSDF sampling instruction streams separate and with different opcodes?

Reply:
yea, this is pre-order traversal creation. Post-order is somewhere else around. In pre-order i'm using 2nd DWORD for jump offset (in post-order 2nd DWORD is used for registers encoding)

> and with different opcodes?

hm? how different?

Reply:
as we've discussed, having certain opcodes around for NEE/LightSampling is pointless, such as the smooth BxDFs

Also Smooth Plastic needs a special code-path which basically boils down to specially modulated diffuse